db/commitlog/commitlog.cc

```diff
@@ -422,6 +460,57 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     seastar::semaphore _reserve_recalculation_guard;
 };
 
+future<> db::commitlog::segment_manager::named_file::open(open_flags flags, file_open_options opt) noexcept {
+    assert(!*this);
+    auto f = co_await open_file_dma(_name, flags, opt);
+    auto existing_size = (flags & open_flags::create) == open_flags{}
+        ? co_await f.size()
+        : 0
+        ;
+    assign(std::move(f));
+    _known_size = existing_size;
+}
+
+future<> db::commitlog::segment_manager::named_file::rename(std::string_view to) {
+    assert(!*this);
+    auto s = sstring(to);
+    co_await seastar::rename_file(_name, s);
+    _name = s;
+}
+
+template <typename CharType>
+future<size_t> db::commitlog::segment_manager::named_file::dma_write(uint64_t pos, const CharType* buffer, size_t len, const io_priority_class& pc, io_intent* intent) noexcept {
+    return file::dma_write(pos, buffer, len, pc, intent).then([this, pos](size_t res) {
+        maybe_update_size(pos + res);
```

Comment:
This just adds an allocation for every write.

Reply:
Better than making it a coroutine frame methinks...

Reply:
Sure, but worse than keeping it in the caller.

Reply:
If it really shows up on perf runs (esp. after other rewrites of write path), it can be changed to return a functor object (non-allocating) that simply increments the size on destructor, so `co_await file.dma_write()(pos, buffer, len)` instead. But seems a bit cumbersome unless proven to really matter. Will do the perf runs again.

Reply:
I have maybe a stupid question:
Why is it important to update the size only **after** the data was written? What do we lose if we update it immediately? If we update it immediately, even before calling `dma_write()`, we no longer have the extra allocation.

Reply:
If we get an exception, we will have updated it to something that is wrong. Could maybe subtract on exception, but that is just as much an attached task as a then...

Reply:
If we get an exception, does the file size even matter any more? Doesn't it mean the entire file is ruined anyway?

Reply:
The file format is built to be somewhat resilient, so no, one failed write is not necessarily the end of the world, though in periodic mode we have no way of back-propagating the info that we mutation write was not backed by a successful CL write. The next buffer write could succeed, and that part of the file at least will be ok. 

Not sure what the best approach is here. My OCD wants to update after the operation. Otoh, we do _pre_ updates for things like write position (because a failure is not terminal). So maybe writes could pre-op update size.

Reply:
But there is one big caveat: dma_write is not 100% guaranteed to fulfill the whole write op in one go IIRC. That is literally why the write path has a big loop popping off its buffer front. Does not happen often, but at least in the olden days I have seen more than one incomplete write.

Reply:
ׁHmm, so if the caller is anyway forced to call this function in a loop, using continuations, it could update the size.
But I'm out of my element here. I'm not familiar with all those code paths. I'll leave it to you to decide (and to Avi to decide if this extra continuation bothers him).

Reply:
It's safer to update post-write, even though it costs a continuation.